ur_robot_driver/src/ros/hardware_interface.cpp

```diff
@@ -628,6 +648,35 @@ void HardwareInterface::publishPose()
   }
 }
 
+void HardwareInterface::extractRobotStatus()
+{
+  // robot status bit 2: Is teach button pressed
+  robot_status_resource_.mode = robot_status_bits_[2] ? RobotMode::MANUAL : RobotMode::AUTO;
```

Comment:
I guess I do have an additional comment: please use some constants here for the bit indices instead of magic nrs.

Reply:
Is this really the right check? The robot can be in manual mode without the teach button being pressed, right?

I am aware, that usually on industrial robots manual and automatic modes have a different meaning than the one used on UR robots. But boiling it down to the teach button pressed feels misleading to me, as well.

Reply:
@fmauch: this was ported from `ur_modern_driver` I believe.

Would you happen to have a suggestion on how to better determine whether the controller is in `AUTO` or `MANUAL`?

I don't even believe URs really have that concept.

When they're being "freedrived" though they are definitely not in `AUTO` (or whatever we want to call the equivalent on URs).

Reply:
> I don't even believe URs really have that concept.

They do (on e-Series, on CB3 the same concept was called "running mode" and "programming mode"), but that concept is different (or a little less restrictive). In `AUTO` mode (which you can activate after setting a password) one can only load and execute predefined programs. (See section "Operational Mode Selection")

However, I am currently not aware of a method receiving this information. (which, of course, doesn't necessarily mean, that it doesn't exist)

> When they're being "freedrived" though they are definitely not in AUTO (or whatever we want to call the equivalent on URs).

Yes, but it feels as if we would be misusing the field from the message for something semantically different. Which, as you reminded me often, not the best thing to do :-)

Reply:
> Would you happen to have a suggestion on how to better determine whether the controller is in AUTO or MANUAL?

Forgot to answer that... To be honest - I don't know. I think, using what UR calls manual and automatic operational mode comes closest to what other industrial robots semantically state with this.

Reply:
> on e-Series, on CB3 the same concept was called "running mode" and "programming mode"

O yes, I'd forgotten about those. I hardly ever use the TP to program anything on these robots, hence why I hadn't thought of it.

> However, I am currently not aware of a method receiving this information. (which, of course, doesn't necessarily mean, that it doesn't exist)

Yes, would be great if we could get that information and use it here, as:

> it feels as if we would be misusing the field from the message for something semantically different. Which, as you reminded me often, not the best thing to do :-)

I completely agree here.

Reply:
I agree with your points, but as you said @gavanderhoorn , this implementation is based on the interpretation of the ur_modern_driver.
I also thought about the issue when porting to the RTDE interface, but I didn't find any possibility to read this information about the Operational Mode over the RTDE interface (see [RTDE Guide](https://www.universal-robots.com/how-tos-and-faqs/how-to/ur-how-tos/real-time-data-exchange-rtde-guide/) )...

Another point to think of: What would we do with the CB-series? I think in the old Polyscope versions, you have to use a configurable input to change to running mode. Yet we could use that as "robot mode" if we have the information.

Reply:
What about limiting this PR to porting the `ur_modern_driver` implementation, then using a potential follow-up PR after/if we get more information on what @fmauch mentions?

It would be a nice enhancement over the partial implementation from `ur_modern_driver`.

@axelschroth: if you could address my comment about magic nrs which started this thread, I'd vote for merging it then.

Reply:
I'd be happy about that, I will commit the changes about the magic nrs tomorrow.

Reply:
> What about limiting this PR to porting the `ur_modern_driver` implementation, then using a potential follow-up PR after/if we get more information on what @fmauch mentions?

I'd be fine with that.

Reply:
I addressed your suggestion using an enum in the rtde client, because this belongs to the rtde specification. Are you okay with that?